man/zabbix_js.man
.\" Start URL macro. Copied from an-ext.tmac for portability
.de UR
.  ds m1 \\$1\"
.  nh
.  if \\n(mH \{\
.    \" Start diversion in a new environment.
.    do ev URL-div
.    do di URL-div
.  \}
..
.\" End URL macro. Copied from an-ext.tmac for portability
.de UE
.  ie \\n(mH \{\
.    br
.    di
.    ev
.
.    \" Has there been one or more input lines for the link text?
.    ie \\n(dn \{\
.      do HTML-NS "<a href=""\\*(m1"">"
.      \" Yes, strip off final newline of diversion and emit it.
.      do chop URL-div
.      do URL-div
\c
.      do HTML-NS </a>
.    \}
.    el \
.      do HTML-NS "<a href=""\\*(m1"">\\*(m1</a>"
\&\\$*\"
.  \}
.  el \
\\*(la\\*(m1\\*(ra\\$*\"
.
.  hy \\n(HY
..
.TH ZABBIX_JS 8 "2019\-01\-29" Zabbix
.SH NAME
zabbix_js \- Zabbix JS utility
.SH SYNOPSIS
.B zabbix_js \-s
.I script\-file
.B \-p
.I input\-param
.RB [ \-l
.IR log\-level ]
.RB [ \-t
.IR timeout ]
.br
.B zabbix_js \-s
.I script\-file
.B \-i
.I input\-file
.RB [ \-l
.IR log\-level ]
.RB [ \-t
.IR timeout ]
.br
.B zabbix_js \-h
.br
.B zabbix_js \-V
.SH DESCRIPTION
.B zabbix_js
is a command line utility that can be used for embedded script testing.
.SH OPTIONS
.IP "\fB\-s\fR, \fB\-\-script\fR \fIscript\-file\fR"
Specify the file name of the script to execute. If '-' is specified as file name, the script will be read from stdin.
.IP "\fB\-p\fR, \fB\-\-param\fR \fIinput\-param\fR"
Specify the input parameter.
.IP "\fB\-i\fR, \fB\-\-input\fR \fIinput\-file\fR"
Specify the file name of the input parameter. If '-' is specified as file name, the input will be read from stdin.
.IP "\fB\-l\fR, \fB\-\-loglevel\fR \fIlog\-level\fR"
Specify the log level.
.IP "\fB\-t\fR, \fB\-\-timeout\fR \fItimeout\fR"
Specify the timeout in seconds.
.IP "\fB\-h\fR, \fB\-\-help\fR"
Display this help and exit.
.IP "\fB\-V\fR, \fB\-\-version\fR"
Output version information and exit.
.SH "EXAMPLES"
\fBzabbix_js \-s script-file.js \-p example \fR
.SH "SEE ALSO"
.UR https://www.zabbix.com/manuals
Documentation
.UE
.PP
.BR zabbix_agentd (8),
.BR zabbix_proxy (8),
.BR zabbix_sender (8),
.BR zabbix_server (8),
.BR zabbix_get (8),
.BR zabbix_agent2 (8)


ChangeLog
Changes for 4.4.5rc1

New features:
..F....... [ZBXNEXT-5473] reworked template linking with Multiselect and loading macro list with AJAX (averza, Ivo, vmurzins)

Bug fixes:
..F....... [ZBX-16481] fixed spelling issues in the code (miks)
.......... [ZBX-1977] reverted fix for axis labels calculation (rlataria)
..F....... [ZBX-16894] fixed dynamic rows losing old input at form error (talbergs)
..F....... [ZBX-17019] fixed aria-live message partially visible beneath multiselect controls (averza)
........S. [ZBX-16966] fixed not operator in correlation function expression (atumilovics)

--------------------------------------------------------------------------------
Changes for 4.4.4

4.4.4rc2 was released as 4.4.4 without any changes

--------------------------------------------------------------------------------
Changes for 4.4.4rc2

Bug fixes:
....I....T [ZBXNEXT-5386] added missing webhook parameters to default database data (Sasha)

--------------------------------------------------------------------------------
Changes for 4.4.4rc1

New features:
.........T [ZBXNEXT-5623] added Pushover webhook integration (vitaly)
.........T [ZBXNEXT-5591] added media "Mattermost" (afayantsev)
.........T [ZBXNEXT-5586] added media "Opsgenie" (mchudinov)
...G...... [ZBXNEXT-5609] added agent2 Windows build support (wiper)
A.F....... [ZBXNEXT-5271] implemented 'delete missing' option for imported template linkages (miks)
.........T [ZBXNEXT-5600] updated templates to internal version v0.34 (vitaly)
.........T [ZBXNEXT-5572] added windows services template, updated windows template objects descriptions, added missing user macro (vitaly)

Bug fixes:
..F....... [ZBX-17101] fixed url validation before output (miks)
...G...... [ZBX-17070] fixed agent2 plugin configuration not being called if no plugin specific parameters are set (wiper)
........S. [ZBX-17062] fixed LLD macro substitution when PostgrSQL is used (viktors)
.......PS. [ZBX-16929] removed redundant code block (mabele)
..F....... [ZBX-16804] added missing get parameters in availability report (rlataria)
..F....... [ZBX-16286] fixed graph widget dimensions errors on high DPI screens (agriscenko)
..F....... [ZBX-16848] fixed fatal error occurring in user profile and user edit forms when PHP FileInfo extension does not exist (Ivo)
..F....... [ZBX-16684] fixed incorrect double quotes in history plain text view (Ivo)
..F....... [ZBX-16918] fixed widget saving with slow internet connection (rlataria)
A......... [ZBX-16903] fixed validation of "interfaceid" field for HTTP agent items with large IDs (Ivo)
..F....... [ZBX-16751] fixed support for php 7.4 (talbergs)
..F....... [ZBX-16879] changed condition description message for tag value in actions and event correlations (vasilijs)
........S. [ZBX-16459] fixed context-aware LLD macro expansion in JsonPath preprocessing (atumilovics)
..F....... [ZBX-16911] fixed multiselect searching suggests in template mass update (rlataria)
.......PS. [ZBX-9084] fixed high memory usage during startup (vso)
...G...... [ZBX-17058] fixed agent2 passive check timemouts exhausting plugin capacity (wiper)
...G...... [ZBX-16932] fixed build fail on netBSD (mabele)
..F....... [ZBX-15041] fixed request not being cancelled along with popup window in widgets (talbergs)
A.F....... [ZBX-16111] removed templateid from screen API output (vasilijs)
..F....... [ZBX-16919] fixed server check warning width in chrome (ashubin, rlataria)
..F....... [ZBX-16954] fixed graph widget scroll disappear (rlataria)
..F....... [ZBX-16995] fixed widget vertically stretch (averza, rlataria)
..F....... [ZBX-16978] fixed undefined offset in problem by severity widget (rlataria)
..F....... [ZBX-16944] fixed "Type of information" field in item form being marked as required when it is read-only (Ivo)
..F....... [ZBX-16912] fixed image map elements having a hand cursor when there is no context menu available (Ivo)
..F....... [ZBX-1977] fixed axis labels calculation (ashubin, rlataria)
..F....... [ZBX-16877] fixed long SNMP OID value goes beyond fields block (rlataria)
..F....... [ZBX-17034] updated Zabbix website links (averza)
..F....... [ZBX-16874] fixed map-type widget clipping in Internet Explorer (averza)
.......PS. [ZBX-16915] fixed memory leak (mabele)
........S. [ZBX-17008] improved Zabbix server performance when using maintenance (vso)
........S. [ZBX-16779] fixed infinite loop when writing export to file fails (mabele)
.........T [ZBX-16540] splitted host_resources templates into 3 subtemplates: for CPU, memory and storage (vitaly)
..F....... [ZBX-16852] fixed graph widget multiselect collapse (ashubin, rlataria)
..F....... [ZBX-16649] fixed flexible textarea behavior (averza, vasilijs)

--------------------------------------------------------------------------------
Changes for 4.4.3

Bug fixes:
...G...PS. [ZBX-16988] fixed agent memory leak (atumilovics)

--------------------------------------------------------------------------------
Changes for 4.4.2

4.4.2rc2 was released as 4.4.2 without any changes

--------------------------------------------------------------------------------
Changes for 4.4.2rc2

Bug fixes:
........S. [ZBX-16955] fixed performance of history syncers and timer processes by not locking each other when suppressing events (vso)

--------------------------------------------------------------------------------
Changes for 4.4.2rc1

New features:
.........T [ZBXNEXT-5493,ZBX-16540] moved LLD rules from parent templates to linked templates for Module HOST-RESOURCES-MIB SNMPv2, Module Interfaces Windows SNMPv2, Net Arista SNMPv2, OS Windows SNMPv2 (vitaly)
...G...... [ZBXNEXT-5570] changed agent2 plugin configuration, moved MaxLinesPerSecond, EnableRemoteCommands, LogRemoteCommands from global to plugin configuration (wiper)
..F....... [ZBX-16871] fixed not setting the default values of multiselects on initial load (averza)
..F....... [ZBX-16776] reworked custom item select to multiselect (rlataria)

src/zabbix_js/zabbix_js.c
/*
** Zabbix
** Copyright (C) 2001-2019 Zabbix SIA
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
**/

#include "common.h"

#include "cfg.h"
#include "log.h"
#include "zbxgetopt.h"
#include "zbxembed.h"
#include "mutexs.h"

const char	*progname;
const char	title_message[] = "zabbix_js";
const char	syslog_app_name[] = "zabbix_js";
const char	*usage_message[] = {
	"-s script-file", "[-i input-file]", "[-p input-param]", NULL,
	"-l log-level", NULL,
	"-h", NULL,
	NULL	/* end of text */
};

unsigned char	program_type;

const char	*help_message[] = {
	"Execute script using Zabbix embedded scripting engine.",
	"",
	"General options:",
	"  -s,--script script-file      Specify the filename of script to execute. Specify - for"
	"                               standard input.",
	"  -i,--input input-file        Specify input parameter file name. Specify - for",
	"                               standard input.",
	"  -p,--param input-param       Specify input parameter",
	"  -l,--loglevel log-level      Specify log level",
	"  -t,--timeout timeout         Specify timeout in seconds",
	"  -h --help                    Display this help message",
	"",
	"Example:",
	"  zabbix_js -s script-file.js -p example",
	NULL	/* end of text */
};


/* long options */
struct zbx_option	longopts[] =
{
	{"script",			1,	NULL,	's'},
	{"input",			1,	NULL,	'i'},
	{"param",			1,	NULL,	'p'},
	{"loglevel",			1,	NULL,	'l'},
	{"timeout",			1,	NULL,	't'},
	{"help",			0,	NULL,	'h'},
	{"version",			0,	NULL,	'V'},
	{NULL}
};

/* short options */
static char	shortopts[] = "s:i:p:hVl:t";

/* end of COMMAND LINE OPTIONS */

unsigned int	configured_tls_connect_mode;
unsigned int	configured_tls_accept_modes;	/* not used in zabbix_get, just for linking */
									/* with tls.c */
char	*CONFIG_TLS_CONNECT		= NULL;
char	*CONFIG_TLS_ACCEPT		= NULL;	/* not used in zabbix_js, just for linking with tls.c */
char	*CONFIG_TLS_CA_FILE		= NULL;
char	*CONFIG_TLS_CRL_FILE		= NULL;
char	*CONFIG_TLS_SERVER_CERT_ISSUER	= NULL;
char	*CONFIG_TLS_SERVER_CERT_SUBJECT	= NULL;
char	*CONFIG_TLS_CERT_FILE		= NULL;
char	*CONFIG_TLS_KEY_FILE		= NULL;
char	*CONFIG_TLS_PSK_IDENTITY	= NULL;
char	*CONFIG_TLS_PSK_FILE		= NULL;

int	CONFIG_PASSIVE_FORKS		= 0;	/* not used in zabbix_js, just for linking with tls.c */
int	CONFIG_ACTIVE_FORKS		= 0;	/* not used in zabbix_js, just for linking with tls.c */

static char	*read_file(const char *filename, char **error)
{
	char	buffer[4096];
	int	n, fd;
	char	*data = NULL;
	size_t	data_alloc = 0, data_offset = 0;

	if (0 != strcmp(filename, "-"))
	{
		if (-1 == (fd = open(filename, O_RDONLY)))
		{
			*error = zbx_strdup(NULL, zbx_strerror(errno));
			return NULL;
		}
	}
	else
		fd = STDIN_FILENO;

	while (0 != (n = read(fd, buffer, sizeof(buffer))))
	{
		if (-1 == n)
		{
			if (fd != STDIN_FILENO)
				close(fd);
			zbx_free(data);
			*error = zbx_strdup(NULL, zbx_strerror(errno));
			return NULL;
		}
		zbx_strncpy_alloc(&data, &data_alloc, &data_offset, buffer, n);
	}

	if (fd != STDIN_FILENO)
		close(fd);

	return data;
}

static char	*execute_script(const char *script, const char *param, int timeout, char **error)
{
	zbx_es_t	es;
	char		*code = NULL;
	int		size;
	char		*errmsg = NULL, *result = NULL;

	zbx_es_init(&es);
	if (FAIL == zbx_es_init_env(&es, &errmsg))
	{
		*error = zbx_dsprintf(NULL, "cannot initialize scripting environment: %s", errmsg);
		return NULL;
	}

	if (0 != timeout)
		zbx_es_set_timeout(&es, timeout);

	if (FAIL == zbx_es_compile(&es, script, &code, &size, &errmsg))
	{
		*error = zbx_dsprintf(NULL, "cannot compile script: %s", errmsg);
		goto out;
	}

	if (FAIL == zbx_es_execute(&es, script, code, size, param, &result, &errmsg))
	{
		*error = zbx_dsprintf(NULL, "cannot execute script: %s", errmsg);
		goto out;
	}
out:
	if (FAIL == zbx_es_destroy_env(&es, &errmsg))
	{
		zbx_error("cannot destroy scripting environment: %s", errmsg);
		zbx_free(result);
	}

	zbx_free(code);
	zbx_free(errmsg);

	return result;
}

int	main(int argc, char **argv)
{
	int	ret = FAIL, loglevel = LOG_LEVEL_WARNING, timeout = 0;
	char	*script_file = NULL, *input_file = NULL, *param = NULL, ch, *script = NULL, *error = NULL,
		*result = NULL;

	progname = get_program_name(argv[0]);

	/* parse the command-line */
	while ((char)EOF != (ch = (char)zbx_getopt_long(argc, argv, shortopts, longopts, NULL)))
	{
		switch (ch)
		{
			case 's':
				if (NULL == script_file)
					script_file = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'i':
				if (NULL == input_file)
					input_file = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'p':
				if (NULL == param)
					param = zbx_strdup(NULL, zbx_optarg);
				break;
			case 'l':
				loglevel = atoi(zbx_optarg);
				break;
			case 't':
				timeout = atoi(zbx_optarg);
				break;
			case 'h':
				help();
				ret = SUCCEED;
				goto clean;
			case 'V':
				version();
				ret = SUCCEED;
				goto clean;
			default:
				usage();
				goto clean;
		}
	}

	if (SUCCEED != zbx_locks_create(&error))
	{
		zbx_error("cannot create locks: %s", error);
		goto clean;
	}

	if (SUCCEED != zabbix_open_log(LOG_TYPE_UNDEFINED, loglevel, NULL, &error))
	{
		zbx_error("cannot open log: %s", error);
		goto clean;
	}


	if (NULL == script_file || (NULL == input_file && NULL == param))
	{
		usage();
		goto close;
	}

	if (NULL != input_file && NULL != param)
	{
		zbx_error("input and script options are mutually exclusive");
		goto close;
	}

	if (0 == strcmp(script_file, "-") && NULL != input_file && 0 == strcmp(input_file, "-"))
	{
		zbx_error("cannot read script and input parameters from standard input at the same time");
		goto close;
	}

	if (NULL == (script = read_file(script_file, &error)))
	{
		if (NULL != error)
			zbx_error("cannot read script file: %s", error);
		else
			zbx_error("cannot use empty script file");

		goto close;
	}

	if (NULL != input_file)
	{
		if (NULL == (param = read_file(input_file, &error)))
		{
			if (NULL != error)
				zbx_error("cannot read input file: %s", error);
			else
				zbx_error("cannot use empty input file");

			goto close;
		}
	}

	if (NULL == (result = execute_script(script, param, timeout, &error)))
	{
		zbx_error("error executing script:\n%s", error);
		goto close;
	}
	ret = SUCCEED;
	printf("\n%s\n", result);
close:
	zabbix_close_log();
clean:
	zbx_free(result);
	zbx_free(error);
	zbx_free(script);
	zbx_free(script_file);
	zbx_free(input_file);
	zbx_free(param);

	return SUCCEED == ret ? EXIT_SUCCESS : EXIT_FAILURE;
}
